### PR TITLE
Feature: PDF actions - merge, split & rotate

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

[feature-merge-tool](https://github.com/paperless-ngx/paperless-ngx/tree/feature-merge-tool) (see https://github.com/jonaswinkler/paperless-ng/pull/789 ) is great and it actually still exists and could be revived, but I think it suffers from the "perfect is the enemy of good" adage. This is much simpler, of course. I guarantee some will want more complex tools but the goal here is to allow light editing, eg split/merge/rotate and I still think we leave the heavier stuff for outside of pngx. 

Besides a code review, Im happy to discuss the merits of either approach, but obviously by opening this PR I'd suggest this is the way forward for now. These can of course be iterated upon.

The only other note here is that rotate actually modifies the original, thats a first.

Screenshot 2024-03-13 at 10 00 36 AM
Screenshot 2024-03-13 at 10 00 49 AM
Screenshot 2024-03-13 at 10 01 19 AM
Screenshot 2024-03-13 at 10 01 26 AM
Screenshot 2024-03-13 at 10 01 53 AM

Closes #367
Closes #396
Closes #505
Closes #2835
Others, Im sure...

## Type of change

- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by andymarden** [comment]
Excellent guys!

**Comment by marcelbrueckner** [comment]
Great feature.

I often have to rotate a single page of a document only. Is this possible?
Or do I have to split it into separate documents, rotate the desired document and merge them all again afterwards?

**Comment by shamoon** [comment]
> I often have to rotate a single page of a document only. Is this possible?

No, not here.

**Comment by hoelter** [comment]
Thanks for doing this -- as a new paperless user who has been interested in some merge functionality, this looks great to me!

**Review by stumpylog — COMMENTED** [review]
I still need to test the functionality itself, but a couple suggestions to think on first

**Comment by shamoon** [comment]
Thanks stumpy, those seem like a good idea, yea. Let me know if you have other thoughts / when you have a chance to test. Definitely no rush from my end on this.

**Comment by wduda** [comment]
I wanted to say that this allows to remove empty pages with a bit of a workaround - split so the white/empty page is a single doc, then delete that whole doc. So with a little effort once can reach that outcome too, and the feature gets an extra thumbs up from me :-)

**Comment by NetRat88** [comment]
I just wanted to get back to work and scan my full inbox (more than 90 documents). Then I looked again for the split function and saw that this feature is almost available. I love it!!! Am I correct that only one review is missing until the feature is rolled out or when can we expect when it will be released? Then I would postpone my work a bit and wait until I scan the documents. Because than I can only put all the docs into my scanner 😍

File: src/documents/bulk_edit.py [hunk 1]
```diff
@@ -146,3 +157,127 @@ def set_permissions(doc_ids, set_permissions, owner=None, merge=False):
...
+            source=DocumentSource.ConsumeFolder,
+            original_file=filepath,
+        ),
+        overrides,
+    )
+
+    return "OK"
+
+
+def split(doc_ids: list[int], pages: list[list[int]]):
```
Comment (stumpylog) [inline src/documents/bulk_edit.py:248]: Probably more important even here, the call to `remove_unreferenced_resources()`, as things will be moving more in the split.  The PDF version shouldn't matter though

**Review by stumpylog — CHANGES_REQUESTED** [review]
Alright, I've been able to do some testing:

- It might be nice if a user tries to split a single page document, to disable the "Add split" button
- `qpdf` doesn't report any errors on the produced split files, that's great
- When rotating, the thumbnail `Last-Modified` header isn't getting reset, due to cache
  - I'll push something up, unless you want to take it on (maybe a general clear cache for document x function for inside `bulk_update_documents`)
  - Same thing probably happens for the other caches

File: src/documents/bulk_edit.py [hunk 2]
```diff
@@ -146,3 +157,127 @@ def set_permissions(doc_ids, set_permissions, owner=None, merge=False):
...
+
+    for doc in qs:
+        try:
+            with pikepdf.open(doc.source_path, allow_overwriting_input=True) as pdf:
+                for page in pdf.pages:
+                    page.rotate(degrees, relative=True)
+                pdf.save()
+                doc.checksum = hashlib.md5(doc.source_file.read()).hexdigest()
+                doc.save()
+                update_document_archive_file.delay(
```
Comment (stumpylog) [inline src/documents/bulk_edit.py:178]: Although unlikely, I'm concerned about a race condition between the 2 async tasks here.

I think a [celery chord or chain or just group](https://docs.celeryq.dev/en/stable/userguide/canvas.html#the-primitives) would prevent that, so these first task(s) are done before the bulk update can happen

**Comment by shamoon** [comment]
Thanks stumpy. Yea I wasnt sure if updating the modified time would be enough but happy to explicitly clear the cache. And good catch about splitting a single page doc and the remove_unreferenced_resources

File: src/documents/bulk_edit.py (hunk 2 above)
Comment (shamoon) [reply src/documents/bulk_edit.py]: Took a stab at this, let me know if thats what you had in mind (or not)

**Comment by stumpylog** [comment]
Alright, gave the latest a pretty good splitting test, 48 page document split at every page.  I got some failures, but that was just due to SQLite limitations (the database being locked).

For rotation though, if I upload an image file, I'm allowed to rotate it, but it will try to open a JPEG as a PDF, for example.  Since the rotation applies to the originals, we should either check the document mime is PDF or directly exclude them from the queryset.  The first option does let us tell the use why it is skipped

**Comment by stumpylog** [comment]
Alternative as well is to check if the mime is an image type, and rotate with PIL instead

**Comment by Keridos** [comment]
Nice to see this. I noticed the conversation about this feature did not include one of my workflows:
I get emails with invoices etc. and I use the email integration to ingest them into paperless. However when. f.ex. multiple pdfs or jpgs are present it always treats them as seperate documents.
So right now I have some documents that only are 1 page long and a part of a multipage document. So far I had no way of merging them together.

**Comment by shamoon** [comment]
Thanks stumpy appreciate all the input / testing. I had incorrectly thought that pikepdf would just fail (thus skip) e.g. an image, I've made it more explicit now. I thought about PIL but I'll leave that as a future enhancement to keep this focused, for now.

Let me know if the latest push addresses that appropriately in your view

**Comment by ccoenen** [comment]
should this latest change (PDF only) also change the UI? If I'm reading this correctly, right now the rotate UI is presented for JPGs as well, it just doesn't do anything. If that's the case, I'd suggest also disabling the interface for JPG based documents (maybe even with a helpful hint that this currently only applies to PDFs)

**Comment by shamoon** [comment]
> I'd suggest also disabling the interface for JPG based documents (maybe even with a helpful hint that this currently only applies to PDFs)

It already does this.

File: src/documents/tests/test_bulk_edit.py [hunk 3]
```diff
@@ -275,3 +278,266 @@ def test_set_permissions_merge(self, m):
             self.doc1,
         )
         self.assertEqual(groups_with_perms.count(), 2)
+
+
+class TestPDFActions(DirectoriesMixin, TestCase):
+    def setUp(self):
+        super().setUp()
+        sample1 = os.path.join(self.dirs.scratch_dir, "sample.pdf")
```
Comment (stumpylog) [inline src/documents/tests/test_bulk_edit.py:286]: Sorry, it's 100% a nit, but this should be possible as a Path operation as well.  I believe I converted the utility to output Path types

**Comment by andymarden** [comment]
Not sure if this has already been covered, but will split and merge new possible if you are not the owner? Deleting docs is only possible if you are the owner even if multiple users have full access and this is essentially deleting a docs as part of the process. That wouldn't be great in terms of: wife scans stuff in and then I do the merging of pages, for example.

File: src/documents/tests/test_bulk_edit.py (hunk 3 above)
Comment (shamoon) [reply src/documents/tests/test_bulk_edit.py]: ha no worries, lmk if that does it.

**Comment by shamoon** [comment]
> Not sure if this has already been covered, but will split and merge new possible if you are not the owner? Deleting docs is only possible if you are the owner even if multiple users have full access and this is essentially deleting a docs as part of the process. That wouldn't be great in terms of: wife scans stuff in and then I do the merging of pages, for example.

no documents are deleted, rotate is the only operation that affects the original so it does require ownership

**Comment by andymarden** [comment]
So if you merge, the original docs stay there and you get one extra? Same with a split, the combined one is still there. The delete is then a separate operation after the operations?

That doesn't seem intuitive, but certainly better than non-owners not being able to split/merge.

I guess a reasonable approach is to tag the original(s) after split/merge as "to be delete after split/merge". That would be good to be configurable via workflow or automatically marked in and way.

But very excited to see this available.

**Review by stumpylog — APPROVED** [review]
Looks good to me.  Much requested feature

**Comment by boeserwolf** [comment]
Thank you for this amazing addition! Would it be possible to add a "Delete Page" option for removing empty or irrelevant pages? 

Unfortunately, a lot of disk space gets wasted for pages like the ones attached:
image

**Comment by shamoon** [comment]
As you can see thats not described above, so the answer is no. Feel free to open a FR.

**Comment by cgraefe** [comment]
Thanks a lot for this great new feature! I'm still quite new to Paperless and I'm impressed by the pace at which it is still moving.

One thing I noticed: I am using Firefox and I had the setting "Use PDF viewer provided by the browser" enabled. This seems to cause the "Split" Menu item to be disabled for all documents. Is this intentional or due to browser limitations? Or has this just slipped by?

**Comment by rsocko** [comment]
I have to say this is a GREAT new addition to the platform and something I've already used several times over. Example: I received an entire history of account invoices in a single, combined PDF. Was able to easily split it up and then process those individual files. Love it!!!

A couple of odd behaviors that I will test a little further to see if I can reproduce: 

- When splitting a 49 page document into 40 or so individual files, the numbering in the list of split items got confused when I was 30+ deep and realized I added a split in a place I shouldn't have. Will try to repro.
- When merging two or more files sometimes the titles didn't appear in the list of items to merge. Again - going to try to repro as it might have been something client-side cached or otherwise.

Altogether this is an AWESOME new feature and super happy to have it.

**Comment by beercity** [comment]
I really appreciate this feature. I had a question though, where do the split documents go before they are processed? I had a case where I split a large document into about 20 smaller ones, and about 4 failed to process due to the database being locked. I am working under the assumption that because my had 2 cores, and I had 2 workers, there were insufficient CPU resources for the database to operate properly. I had looked in /tmp and could not find the 4 documents that failed to process. I had to end up splitting the whole document again. I gave the VM an extra core but left the 1 core/worker setting intact, so the third core would be allowed for other operations in the VM, and had no further issues. But I thought it would be good to know where the docs go if they fail to process.

**Comment by shamoon** [comment]
Appreciate all the thoughts and nice words, for questions related to this feature please open a discussion, commenting on a closed PR is not the best way to get help. As such we’ll lock this as resolved. Thanks for your understanding

**Pull request merged**